components/net/http_cache.rs

```diff
@@ -624,6 +624,18 @@ impl HttpCache {
         let entry_key = CacheKey::new(request.clone());
         if let Some(cached_resources) = self.entries.get_mut(&entry_key) {
             for cached_resource in cached_resources.iter_mut() {
+                // done_chan will have been set to Some(..) by http_network_fetch.
+                // If the body is not receiving data, set the done_chan back to None.
+                // Otherwise, create a new channel to let
+                // The response constructed here will replace the 304 one from the network.
+                match *cached_resource.body.lock().unwrap() {
+                    ResponseBody::Receiving(..) => {
+                        let (done_sender, done_receiver) = channel();
+                        *done_chan = Some((done_sender.clone(), done_receiver));
+                        cached_resource.awaiting_body.lock().unwrap().push(done_sender);
```

Comment:
Is there a risk of deadlock here while the cached_resource.body lock is held? Would it be safer to do something like:
```rust
let in_progress_channel = match *cached_resource.body.lock().unwrap() {
    ResponseBody::Receiving(..) => Some(channel())
    ResponseBody::Empty | ResponseBody::Done(..) => None
};
match in_progress_channel {
    Some((done_sender, done_receiver)) => {
        cached_resource.awaiting_body.lock().unwrap().push(done_sender.clone());
        *done_chan = Some((done_sender, done_receiver));
    }
    None => *done_chan = None,
}
```